protocol_driver_test_fixture.cc

```diff
@@ -0,0 +1,119 @@
+// limitations under the License.
+
+#include "protocol_driver_test_fixture.h"
+
+#include "protocol_driver_allocator.h"
+#include "absl/base/internal/sysinfo.h"
+#include "distbench_utils.h"
+#include "glog/logging.h"
+#include "google/protobuf/repeated_field.h"
+
+namespace distbench {
+
+// ProtocolDriver ===============================================
+ProtocolDriverTestFixture::ProtocolDriverTestFixture(int tree_depth) {
+  tree_depth_ = tree_depth;
+}
+
+absl::Status ProtocolDriverTestFixture::Initialize(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+
+  auto pdo = pd_opts;
+  auto server_settings = pdo.mutable_server_settings();
+  google::protobuf::RepeatedPtrField<NamedSetting>::iterator
+      next_protocol_driver_it = server_settings->end();
+  for (auto it=server_settings->begin(); it!=server_settings->end(); it++) {
+      if (it->name()=="driver_under_test") {
+        next_protocol_driver_it = it;
+        pdo.set_protocol_name(it->string_value());
+      }
+  }
+  if (next_protocol_driver_it != server_settings->end())
+    server_settings->erase(next_protocol_driver_it);
+
+  auto maybe_pd_instance_ = AllocateProtocolDriver(pdo, &port_, tree_depth_+1);
+  if (!maybe_pd_instance_.ok()) return maybe_pd_instance_.status();
+  pd_instance_ = std::move(maybe_pd_instance_.value());
+
+  return absl::OkStatus();
+}
+
+absl::Status ProtocolDriverTestFixture::InitializeClient(
+    const ProtocolDriverOptions& pd_opts) {
+  return absl::UnimplementedError("InitializeClient is not implemented for test_fixture.");
+}
+
+absl::Status ProtocolDriverTestFixture::InitializeServer(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+  return absl::UnimplementedError("InitializeServer is not implemented for test_fixture.");
+}
+
+void ProtocolDriverTestFixture::SetHandler(
+    std::function<std::function<void()>(ServerRpcState* state)> handler) {
+  auto server_handler = [=] (ServerRpcState* state) {
+    ++server_rpc_cnt;
+    return handler(state);
+  };
+  pd_instance_->SetHandler(server_handler);
+}
+
+void ProtocolDriverTestFixture::SetNumPeers(int num_peers) {
+  pd_instance_->SetNumPeers(num_peers);
+}
+
+ProtocolDriverTestFixture::~ProtocolDriverTestFixture() {
+}
+
+absl::StatusOr<std::string> ProtocolDriverTestFixture::HandlePreConnect(
+    std::string_view remote_connection_info, int peer) {
+  return pd_instance_->HandlePreConnect(remote_connection_info, peer);
+}
+
+absl::Status ProtocolDriverTestFixture::HandleConnect(
+    std::string remote_connection_info, int peer) {
+  return pd_instance_->HandleConnect(remote_connection_info, peer);
+}
+
+void ProtocolDriverTestFixture::HandleConnectFailure(
+    std::string_view local_connection_info) {
+  pd_instance_->HandleConnectFailure(local_connection_info);
+}
+
+std::vector<TransportStat> ProtocolDriverTestFixture::GetTransportStats() {
+  std::vector<TransportStat> transport_stats;
+  transport_stats.push_back({"client_rpc_cnt", client_rpc_cnt});
```

Comment:
you should be calling the method of the pd below you as well.

Reply:
Done